qiskit_nature/second_q/hamiltonians/lattices/hexagonal_lattice.py

```diff
@@ -0,0 +1,73 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021, 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The hexagonal lattice"""
+from rustworkx import generators
+
+from .lattice import Lattice
+
+
+class HexagonalLattice(Lattice):
+    """Hexagonal lattice."""
+
+    def __init__(
+        self,
+        rows: int,
+        cols: int,
+        edge_parameter: complex = 1.0,
+        onsite_parameter: complex = 0.0,
+    ) -> None:
+        """
+        Args:
+            rows: Length of the x direction.
+            cols: Length of the y direction.
+            edge_parameter: Weight on all the edges, specified as a single value.
+                Defaults to 1.0.
+            onsite_parameter: Weight on the self-loops, which are edges connecting a node to itself.
+                Defaults to 0.0.
+        """
+        self._rows = rows
+        self._cols = cols
+        self._edge_parameter = edge_parameter
+        self._onsite_parameter = onsite_parameter
+
+        graph = generators.hexagonal_lattice_graph(rows, cols, multigraph=False)
+
+        edges = graph.weighted_edge_list()
+
+        # Add edge weights
+        for idx, _ in enumerate(edges):
+            graph.update_edge_by_index(idx, self._edge_parameter)
```

Comment:
It would be slightly more performant to do this as:

```suggestion
        # Add edge weights
        for idx in graph.edge_indices():
            graph.update_edge_by_index(idx, self._edge_parameter)
```

as the iterator only has to convert a single int instead of a tuple and None at each step. You probably wouldn't really see a difference in practice though except at really large graphs though.

It might also be worth benchmarking this against using `range(graph.num_edges())` and `range(graph.num_nodes())` (for the loop below). While assuming a contiguous range of indices for nodes and edges in rustworkx isn't always a good idea (because there can be holes on removals) using this from the fresh output of the generator function should work fine as there will be no removals. This might be potentially even faster because the biggest overhead in rustworkx comes from the rust->python conversion and with any of the other methods listed we do that on each step in the iterator for these loops.

Reply:
yeah just doing some quick benchmarking it does seem to be faster, especially getting to larger lattices:

using `graph.edge_indices()` & `graph.node_indices()`:
- `HexagonalLattice(10,10)`: 4.668223667999996s
- `HexagonalLattice(20,20)`: 18.730463289s
- `HexagonalLattice(25,25)`: 28.666557584000003s

using `range(graph.num_edges())` & `graph.node_indices()`:
- `HexagonalLattice(10,10)`: 4.253540048s
- `HexagonalLattice(20,20)`: 16.220453660000004s
- `HexagonalLattice(25,25)`: 24.83831334300001s

using `range(graph.num_edges())` & `range(graph.num_nodes())`:
- `HexagonalLattice(10,10)`: 3.9827502070000005
- `HexagonalLattice(20,20)`: 15.120913722000004
- `HexagonalLattice(25,25)`: 23.106167259000003